pkg/gpu/probe.go

```diff
@@ -222,16 +199,91 @@ func (p *Probe) cleanupFinished() {
 	p.consumer.cleanFinishedHandlers()
 }
 
-func (p *Probe) startEventConsumer() {
-	handler := ddebpf.NewRingBufferHandler(consumerChannelSize)
+// toPowerOf2 converts a number to its nearest power of 2
+func toPowerOf2(x int) int {
+	log := math.Log2(float64(x))
+	return int(math.Pow(2, math.Round(log)))
+}
+
+// setupSharedBuffer sets up the ringbuffer to handle CUDA events produces by ebpf uprobes
+// it must be called BEFORE the InitWithOptions method of the manager is called
+func setupSharedBuffer(m *manager.Manager, o *manager.Options) {
+	rbHandler := ddebpf.NewRingBufferHandler(consumerChannelSize)
 	rb := &manager.RingBuffer{
-		Map: manager.Map{Name: cudaEventMap},
+		Map: manager.Map{Name: cudaEventsMap},
 		RingBufferOptions: manager.RingBufferOptions{
-			RecordHandler: handler.RecordHandler,
-			RecordGetter:  handler.RecordGetter,
+			RecordHandler: rbHandler.RecordHandler,
+			RecordGetter:  rbHandler.RecordGetter,
 		},
 	}
-	p.mgr.RingBuffers = append(p.mgr.RingBuffers, rb)
-	p.consumer = newCudaEventConsumer(p.sysCtx, handler, p.cfg)
-	p.consumer.Start()
+
+	ringBufferSize := toPowerOf2(defaultRingBufferSize)
+
+	o.MapSpecEditors[cudaEventsMap] = manager.MapSpecEditor{
+		Type:       ebpf.RingBuf,
+		MaxEntries: uint32(ringBufferSize),
+		KeySize:    0,
+		ValueSize:  0,
+		EditorFlag: manager.EditType | manager.EditMaxEntries | manager.EditKeyValue,
+	}
+
+	m.RingBuffers = append(m.RingBuffers, rb)
+	eventHandler = rbHandler
+}
+
+func getAttacherConfig(cfg *config.Config) uprobes.AttacherConfig {
+	return uprobes.AttacherConfig{
+		Rules: []*uprobes.AttachRule{
+			{
+				LibraryNameRegex: regexp.MustCompile(`libcudart\.so`),
+				Targets:          uprobes.AttachToExecutable | uprobes.AttachToSharedLibraries,
+				ProbesSelector: []manager.ProbesSelector{
+					&manager.AllOf{
+						Selectors: []manager.ProbesSelector{
+							&manager.ProbeSelector{ProbeIdentificationPair: manager.ProbeIdentificationPair{EBPFFuncName: cudaLaunchKernelProbe}},
+							&manager.ProbeSelector{ProbeIdentificationPair: manager.ProbeIdentificationPair{EBPFFuncName: cudaMallocProbe}},
+							&manager.ProbeSelector{ProbeIdentificationPair: manager.ProbeIdentificationPair{EBPFFuncName: cudaMallocRetProbe}},
+							&manager.ProbeSelector{ProbeIdentificationPair: manager.ProbeIdentificationPair{EBPFFuncName: cudaStreamSyncProbe}},
+							&manager.ProbeSelector{ProbeIdentificationPair: manager.ProbeIdentificationPair{EBPFFuncName: cudaStreamSyncRetProbe}},
+							&manager.ProbeSelector{ProbeIdentificationPair: manager.ProbeIdentificationPair{EBPFFuncName: cudaFreeProbe}},
+						},
+					},
+				},
+			},
+		},
+		EbpfConfig:         &cfg.Config,
+		PerformInitialScan: cfg.InitialProcessSync,
+	}
+}
+
+func getManager(buf io.ReaderAt, opts manager.Options) (*ddebpf.Manager, error) {
```

Comment:
I think if both `getManager` and `setupSharedBuffer` were part of the probe we could avoid the need for a global variable for `eventHandler`, and it would become a field of `Probe`.  As part of that, we should move the `eventHandler.Stop()` call from `cudaEventConsumer.Stop` to `Probe.Close`, so that both start/init is handled by the probe itself.